spec.html

```diff
@@ -23991,16 +24185,277 @@ <h1>Runtime Semantics: Evaluation</h1>
+      <h1>
+        ApplyDecoratorsToElementDefinition (
+          _homeObject_: an Object,
+          _elementRecord_: a ClassElementDefinition Record,
+          _extraInitializers_: a List of function objects,
+          _isStatic_: a Boolean,
+        ): either a normal completion containing ~unused~, or an abrupt completion
+      </h1>
+      <dl class="header">
+        <dt>description</dt>
+        <dd>Applies the decorators stored on an ClassElementDefinition to the value of the definition.</dd>
+      </dl>
+      <emu-alg>
+        1. Let _decorators_ be _elementRecord_.[[Decorators]].
+        1. If _decorators_ is ~empty~, return ~unused~.
+        1. Let _key_ be _elementRecord_.[[Key]].
+        1. Let _kind_ be _elementRecord_.[[Kind]].
+        1. For each element _decoratorRecord_ of _decorators_, do
+          1. Let _decorator_ be _decoratorRecord_.[[Decorator]].
+          1. Let _decoratorReceiver_ be _decoratorRecord_.[[Receiver]].
+          1. Let _decorationState_ be the Record { [[Finished]]: *false* }.
+          1. Let _context_ be CreateDecoratorContextObject(_kind_, _key_, _extraInitializers_, _decorationState_, _isStatic_).
+          1. Let _value_ be *undefined*.
+          1. If _kind_ is ~method~, set _value_ to _elementRecord_.[[Value]].
+          1. Else if _kind_ is ~getter~, set _value_ to _elementRecord_.[[Get]].
+          1. Else if _kind_ is ~setter~, set _value_ to _elementRecord_.[[Set]].
+          1. Else if _kind_ is ~accessor~, then
+            1. Set _value_ to OrdinaryObjectCreate(%Object.prototype%).
+            1. Perform ! CreateDataPropertyOrThrow(_value_, *"get"*, _elementRecord_.[[Get]]).
+            1. Perform ! CreateDataPropertyOrThrow(_value_, *"set"*, _elementRecord_.[[Set]]).
+          1. Let _newValue_ be ? Call(_decorator_, _decoratorReceiver_, « _value_, _context_ »).
+          1. Set _decorationState_.[[Finished]] to *true*.
+          1. If _kind_ is ~field~, then
+            1. If IsCallable(_newValue_) is *true*, append _newValue_ to _elementRecord_.[[Initializers]].
+            1. Else if _newValue_ is not *undefined*, throw a *TypeError* exception.
+          1. Else if _kind_ is ~accessor~, then
+            1. If _newValue_ is an Object, then
+              1. Let _newGetter_ be ? Get(_newValue_, *"get"*).
+              1. If IsCallable(_newGetter_) is *true*, set _elementRecord_.[[Get]] to _newGetter_.
+              1. Else if _newGetter_ is not *undefined*, throw a *TypeError* exception.
+              1. Let _newSetter_ be ? Get(_newValue_, *"set"*).
+              1. If IsCallable(_newSetter_) is *true*, set _elementRecord_.[[Set]] to _newSetter_.
+              1. Else if _newSetter_ is not *undefined*, throw a *TypeError* exception.
+              1. Let _initializer_ be ? Get(_newValue_, *"init"*).
+              1. If IsCallable(_initializer_) is *true*, append _initializer_ to _elementRecord_.[[Initializers]].
```

Comment:
I [opened an issue](https://github.com/babel/babel/issues/15608) against `@babel/plugin-proposal-decorators` thinking they had implemented this spec incorrectly. They in turned pointed me to this line saying that they were indeed implementing the spec appropriately. As you can see from the issue description, appending initializers results in an inside-out evaluation of the decorator's initializers, while everything else is evaluated outside-in (top->bottom, left-right).

Shouldn't initializers be prepended onto `[[Initializers]]` in order to maintain the same outside-in execution order?

Reply:
@pgoforth this was intentional. The ordering is from innermost decorator to outermost. This means innermost initializers run before outermost, and wrapped methods are wrapped from innermost to outermost, with the outermost method becoming the final method on the class. I commented on the other issue illustrating how the ordering of wrapped methods is actually not really predetermined - the outer method chooses when to call the inner. But the wrapping is consistent with the other orderings.

Reply:
@pzuraq Your example will give you the correct logging, but in a practical example the ordering is problematic and does not adhere to the expectations of the spec.

I have created a minimum repro with the issue in question [here](https://github.com/pgoforth/js-decorator-ordering-tests). To sum up, you have to create an overly complex decorator that inverts the order of the setter calls in order to maintain the original order of the initializer calls. I hope you can appreciate how confusing it is having a simple decorator giving an incorrect result because of innermost -> outermost execution.

Reply:
Decorators are *evaluated* left to right top to bottom, but they are *applied* innermost to outermost, and that's certainly what I expect.

Reply:
@ljharb I think I'm not being clear. I have no problem with the ordering, that's irrelevant to my complaint. If the setters were applied innermost to outermost, why wouldn't the evalutation from left to right top to bottom provide the inner setter with the reference to the outer setter? Currently the outer setter can only call in the opposite direction that the initializers are applied in.

```js
function appendString(strToAppend) {
    return function ({ set }) {
        return {
            set(value) {
                // outtermost call is applied first first
                return set.call(this, `${value}${strToAppend}`);
            },
            init(initialValue) {
                // innermost call is applied first first
                return `${initialValue}${strToAppend}`
            },
        }
    }
}

class FakeClass{
    @appendString('Baz')
    @appendString('Bar')
    accessor fakeField = 'foo'
}
const instance = new FakeClassSimple()
console.log(instance.fakeField) // 'fooBarBaz'
// Setting to the same value it's initialized to, results in the opposite result
instance.fakeField = 'foo'
console.log(instance.fakeField) // 'fooBazBar'
```

Can you see how this is confusing?

Reply:
This convo has been resolved with a change in the init order in the spec